src/drivers/ins/microstrain/MicroStrain.cpp

```diff
@@ -0,0 +1,623 @@
+	    && _param_ms_imu_rate_hz.get() > 0) {
+		imu_descriptors[num_imu_descriptors++] = mip_descriptor_rate { MIP_DATA_DESC_SENSOR_GYRO_SCALED, imu_decimation};
+	}
+
+	if (supports_descriptor(MIP_SENSOR_DATA_DESC_SET, MIP_DATA_DESC_SENSOR_MAG_SCALED)
+	    && _param_ms_mag_rate_hz.get() > 0) {
+		imu_descriptors[num_imu_descriptors++] = mip_descriptor_rate { MIP_DATA_DESC_SENSOR_MAG_SCALED, mag_decimation};
+	}
+
+	if (supports_descriptor(MIP_SENSOR_DATA_DESC_SET, MIP_DATA_DESC_SENSOR_PRESSURE_SCALED)
+	    && _param_ms_baro_rate_hz.get() > 0) {
+		imu_descriptors[num_imu_descriptors++] = mip_descriptor_rate { MIP_DATA_DESC_SENSOR_PRESSURE_SCALED, baro_decimation};
+	}
+
+
+	// // Write the settings
+	res = mip_3dm_write_message_format(&device, MIP_SENSOR_DATA_DESC_SET, num_imu_descriptors,
+					   imu_descriptors);
+
+	return res;
+
+}
+
+void MicroStrain::initialize_ins()
+{
+	if (_is_initialized) {
+		return;
+	}
+
+	// first try default baudrate
+	const uint32_t DEFAULT_BAUDRATE = 115200;
+	const uint32_t DESIRED_BAUDRATE = 921600;
+
+	if (connect_at_baud(DEFAULT_BAUDRATE) == PX4_ERROR) {
+
+		static constexpr uint32_t BAUDRATES[] {9600, 19200, 38400, 57600, 115200, 128000, 230400, 460800, 921600};
+		bool is_connected = false;
+
+		for (auto &baudrate : BAUDRATES) {
+			if (connect_at_baud(baudrate) == PX4_OK) {
+				PX4_INFO("found baudrate %" PRIu32, baudrate);
+				is_connected = true;
+				break;
+			}
+		}
+
+		if (!is_connected) {
+			_is_init_failed = true;
+			PX4_WARN("Could not connect to the device, exiting");
+			return;
+		}
+	}
+
+
+	PX4_INFO("mip_base_set_idle");
+
+	if (mip_base_set_idle(&device) != MIP_ACK_OK) {
+		PX4_ERR("ERROR: Could not set the device to idle!");
+		return;
+	}
+
+	PX4_INFO("Setting the baud to desired baud rate");
+
+	usleep(500_ms);
+
+	if (mip_3dm_write_uart_baudrate(&device, DESIRED_BAUDRATE) != MIP_ACK_OK) {
+		PX4_ERR("ERROR: Could not set the baudrate!");
+		_is_init_failed = true;
+		return;
+	}
+
+	tcflush(device_uart.uart_get_fd(), TCIOFLUSH);
+
+	usleep(500_ms);
+
+	for (int i = 0; i < 10; i++) {
+		PX4_INFO("Connection Attempt: %d", i);
+
+		if (connect_at_baud(DESIRED_BAUDRATE) == PX4_OK) {
+			break;
+		}
+
+		if (i >= 9) {
+			PX4_ERR("ERROR: Could not reconnect at desired baud!");
+			_is_init_failed = true;
+			return;
+		}
+	}
+
+	if (get_supported_descriptors() != MIP_ACK_OK) {
+		PX4_INFO("ERROR: Could not get descriptors");
+		_is_init_failed = true;
+		return;
+	}
+
+
+	if (configure_imu_message_format() != MIP_ACK_OK) {
+		PX4_INFO("ERROR: Could not write message format");
+		_is_init_failed = true;
+		return;
+	}
+
+
+	//
+	// Register data callbacks
+	//
+	mip_interface_register_packet_callback(&device, &sensor_data_handler, MIP_SENSOR_DATA_DESC_SET, false, &sensor_callback,
+					       this);
+
+
+	//
+	// Setup the rotation based on PX4 standard rotation sets
+	//
+
+	if (mip_3dm_write_sensor_2_vehicle_transform_euler(&device, math::radians<float>(rot_lookup[rotation].roll),
```

Comment:
Should the sensor to vehicle transform be written to the device?  Is this "double applying" the rotation?

Reply:
removed